From 1e4078efc20d369d08838316aec7b9fab94c39a9 Mon Sep 17 00:00:00 2001
From: Haihao Xiang <haihao.xiang@intel.com>
Date: Sun, 18 Jul 2021 00:51:04 +0800
Subject: [PATCH] msdk: make sure child context is destroyed first

The parent context shares some resources with child context, so the
child context should be destroyed first, otherwise the command below
will trigger a segmentation fault

$> gst-launch-1.0 videotestsrc num-buffers=100 ! msdkh264enc ! \
msdkh264dec ! fakesink videotestsrc num-buffers=50 ! \
msdkh264enc ! msdkh264dec ! fakesink
---
 sys/msdk/gstmsdkcontext.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/sys/msdk/gstmsdkcontext.c b/sys/msdk/gstmsdkcontext.c
index 5e9821c9a..f1c239825 100644
--- a/sys/msdk/gstmsdkcontext.c
+++ b/sys/msdk/gstmsdkcontext.c
@@ -53,6 +53,7 @@ struct _GstMsdkContextPrivate
   gint shared_async_depth;
   GMutex mutex;
   GList *child_session_list;
+  GstMsdkContext *parent_context;
 #ifndef _WIN32
   gint fd;
   VADisplay dpy;
@@ -283,6 +284,8 @@ gst_msdk_context_finalize (GObject * obj)
 #endif
 
 done:
+  if (priv->parent_context)
+    gst_object_unref (priv->parent_context);
   G_OBJECT_CLASS (parent_class)->finalize (obj);
 }
 
@@ -398,6 +401,7 @@ gst_msdk_context_new_with_parent (GstMsdkContext * parent)
   priv->dpy = parent_priv->dpy;
   priv->fd = parent_priv->fd;
 #endif
+  priv->parent_context = gst_object_ref (parent);
 
   return obj;
 }
-- 
2.17.1

